jest/hasteImpl.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+'use strict';
+
+const path = require('path');
+
+const ROOT = path.join(__dirname, '..');
+
+const BLACKLISTED_PATTERNS/*: Array<RegExp>*/ = [
+  /.*\/__(mocks|tests)__\/.*/,
+  /^Libraries\/Animated\/src\/polyfills\/.*/,
+];
+
+const WHITELISTED_PREFIXES/*: Array<string>*/ = [
+  'IntegrationTests',
+  'Libraries',
+  'ReactAndroid',
+  'RNTester',
+];
+
+const NAME_REDUCERS/*: Array<[RegExp, string]>*/ = [
+  // extract basename
+  [/^(?:.*\/)?([a-zA-Z0-9$_.-]+)$/, '$1'],
+  // strip .js/.js.flow suffix
+  [/^(.*)\.js(\.flow)?$/, '$1'],
+  // strip .android/.ios/.native/.web suffix
+  [/^(.*)\.(android|ios|native|web)$/, '$1'],
+];
+
+const haste = {
+  /*
+   * @return {string|void} hasteName for module at filePath; or undefined if
+   *                       filePath is not a haste module
+   */
+  getHasteName(
+    filePath/*: string*/,
+    sourceCode/* : ?string*/
+  )/*: (string | void)*/ {
+    if (!isHastePath(filePath)) {
+      return undefined;
+    }
+
+    const hasteName = NAME_REDUCERS.reduce(
+      (name, [pattern, replacement]) => name.replace(pattern, replacement),
+      filePath
+    );
+
+    return hasteName;
+  },
+};
+
+function isHastePath(filePath/*: string*/)/*: bool*/ {
+  if (!filePath.endsWith('.js') && !filePath.endsWith('.js.flow')) {
+    return false;
+  }
+
+  if (!filePath.startsWith(ROOT)) {
```

Comment:
hey @rubennorte e2e tests are failing right now and my guess is this might be one cause. Refer: https://circleci.com/gh/facebook/react-native/42132?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link.

I'm seeing `ReferenceError: SHA-1 for file /private/tmp/react-native-9ke7hCEE/EndToEndTest/node_modules/fbjs/lib/invariant.js is not computed`. I might be completely wrong but looks like we don't handle haste modules in the node_modules directory?

Reply:
Hi @chirag04, I don't think that specific error is caused by the @providesModule change, but a bad integration between Metro and Watchman. @mjesun or @rafeca can you please confirm that?

Reply:
Hi, yes, that might be. `jest-haste-map` is not computing SHA-1s for `node_modules`; I'm going to put up a fix for this.

Reply:
Hi @mjesun! When might that fix be ready? It's keeping us from running Android tests.